### PR TITLE
Combine semantically equal catch blocks with same var declarations

### PR DESCRIPTION
**Pull request by Marklinzi**
## What's changed?

Fix NPE in CombineSemanticallyEqualCatchBlocks occurring when semantically equal catch blocks contain a variable declaration.

## What's your motivation?

Fixes https://github.com/openrewrite/rewrite-static-analysis/issues/278.

## Have you considered any alternatives or workarounds?

No.

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: src/main/java/org/openrewrite/staticanalysis/CombineSemanticallyEqualCatchBlocks.java [hunk 1]
```diff
@@ -1692,8 +1692,7 @@ public J.VariableDeclarations visitVariableDeclarations(J.VariableDeclarations m
                             nullMissMatch(multiVariable.getTypeExpression(), compareTo.getTypeExpression()) ||
                             multiVariable.getVariables().size() != compareTo.getVariables().size() ||
                             multiVariable.getLeadingAnnotations().size() != compareTo.getLeadingAnnotations().size() ||
-                            doesNotContainSameComments(multiVariable.getPrefix(), compareTo.getPrefix()) ||
-                            doesNotContainSameComments(multiVariable.getVarargs(), compareTo.getVarargs())) {
+                            doesNotContainSameComments(multiVariable.getPrefix(), compareTo.getPrefix())) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/staticanalysis/CombineSemanticallyEqualCatchBlocks.java:1695]: Reading the change set and the recipe I'm wondering what made you make this change here as opposed to a nullsafe check in the `doesNotContainSameComments`. Any context towards that? At first glance it seems we now convert fewer cases even if they could be equal; was that intended or a misreading on my part?

**Review by timtebeek — COMMENTED** [review]
Thanks a lot for covering this case with a test and working out what makes that specific case pass. A quick question as you probably have more context than I do, as I wonder about what led to this specific code change as compared to alternatives.

File: src/main/java/org/openrewrite/staticanalysis/CombineSemanticallyEqualCatchBlocks.java (hunk 1 above)
Comment (Marklinzi) [reply src/main/java/org/openrewrite/staticanalysis/CombineSemanticallyEqualCatchBlocks.java]: Hi, thanks for the review.
I have now reverted the change and made a nullsafe check in `doesNotContainSameComments`.

To be honest, I thought that the check for varargs comments is unnecessary here as we are comparing catch blocks, and I didn't think about features like anonymous classes (or similar constructions) where varargs declarations could occur.
That said, I am still not sure if there is a case where `multiVariable.getVarargs()` is not null, as the `CommentVisitor` is only used when two catch blocks are semantically equal as defined by `SemanticallyEqual.areEqual(...)` and for two anonymous class definitions, this seems to never be the case.

For example, the following two catch blocks could be combined:
``` JAVA
class Test {
    void method() {
        try {
        } catch (A ex) {
            String s = "foo";
            
            new Object() {
              public void test(String... test) {
              }
            };
        } catch (B ex) {
            String s = "foo";
            
            new Object() {
              public void test(String... test) {
              }
            };
        }
    }
}
```

However, this is currently not done.

**Comment by timtebeek** [comment]
Thanks a lot for the detailed look and response to my question. Good to have this case fixed. We expect to follow up with a release early next week. Until then you can try it out from [our snapshot versions](https://docs.openrewrite.org/reference/snapshot-instructions)

**Pull request merged**